network/shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockResolver.java

```diff
@@ -252,4 +294,18 @@ public String toString() {
         .toString();
     }
   }
+
+  /**
+   * write out the set of registered executors to a file so we can reload them on restart.
+   * You must have a lock on executors when calling this
+   */
+  private void saveRegisteredExecutors() throws IOException {
+    if (db != null) {
+      ByteArrayOutputStream bytes = new ByteArrayOutputStream();
+      ObjectOutputStream out = new ObjectOutputStream(bytes);
+      out.writeObject(executors);
+      out.close();
+      db.put("registeredExecutors".getBytes(Charsets.UTF_8), bytes.toByteArray());
```

Comment:
my previous comments got closed based on the change to move opening up out, so just repasting it here so we don't lose it:

I expected to see one entry per application rather then putting them all in.
where are we removing applications?
I would have thought from applicationRemoved we would remove it and save again?